R/taxonomy.R

```diff
@@ -197,6 +209,68 @@ setMethod("checkTaxonomy", signature = c(x = "SummarizedExperiment"),
     }
 )
 
+#' @rdname taxonomy-methods
+#' @importFrom utils assignInMyNamespace
+#' @aliases checkTaxonomy
+#' @export
+# Function to set taxonomy ranks
+setTaxonomyRanks <- function(ranks) {
+    ranks <- tolower(ranks)
+    # Check if rank is a character vector with length >= 1
+    if (!is.character(ranks) || length(ranks) < 1 
+        || any(ranks == "" | ranks == " " | ranks == "\t" | ranks == "-" | ranks == "_")
+        || any(grepl("\\s{2,}", ranks))) {
+        stop("Input 'rank' should be a character vector with non-empty strings,
+             no spaces, tabs, hyphens, underscores, and non-continuous spaces."
+             , call. = FALSE)
+    }
+    
+    assignInMyNamespace("TAXONOMY_RANKS", ranks)
+}
+
+
+#' @rdname taxonomy-methods
+#'@export
+# Function to get taxonomy ranks
+getTaxonomyRanks <- function() {
+    return(TAXONOMY_RANKS)
+}
+
+.set_taxonomy_ranks <- function(tse, ranks, set.ranks = FALSE) {
+    # Check if ranks are not specified by the user
+    if (is.null(ranks)) {
+        # Attempt to retrieve ranks from the column names of rowData(tse)
+        if (!is.null(rowData(tse))) {
+            ranks <- colnames(rowData(tse))
+            if (length(ranks) == 0) {
+                stop("rowData of the 'tse' object has no column names to infer ranks.", call. = FALSE)
+            }
+        } else {
+            stop("rowData of the 'tse' object does not exist or does not have column names.", call. = FALSE)
+        }
+    } else {
+        # User has specified ranks, validate them
+        if (!is.character(ranks) || length(ranks) == 0) {
+            stop("Provided 'ranks' must be a non-empty character vector.", call. = FALSE)
+        }
+    }
```

Comment:
This is little bit too complex with multiple level if statements. Instead you could divide it:

1. Input check. Check that ranks is character vector and set.ranks is logical scalar.
2. If ranks is NULL, get ranks from `colnames(rowData(tse))` (I think rowData cannot be NULL). However, get only those column names that are character. At least QIIME2 inputs some confident numbers.
3. If length of ranks is 0, give error if set.ranks = TRUE. (We must allow user to import TreeSE with empty rowData)